### ISSUE TITLE
Tracking issue for RFC 2196, "metabuild: semantic build scripts for Cargo"

### ISSUE THREAD

**Issue by Centril** [rust-lang/cargo#14903]
This is a tracking issue for the RFC "metabuild: semantic build scripts for Cargo" (rust-lang/rfcs#2196).

**Steps:**

- [x] Implement the RFC (#5628)
- [ ] Implement Multiple Build Scripts
  - [x] Parsing #15630
  - [x] Running multiple build scripts #15704 
- [ ] Implement Build Script Parameter Passing
- [ ] Implement Build Script Delegation
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

- `OUT_DIR`: how to read `OUT_DIR` when there are multiple build scripts was deferred out of #15704
- Conflicts between build scripts: if two build scripts write the same env variable, etc, what should happen?  As of #15704, the last one alphabetically wins (except sometimes the first alphabetically wins).  We need this consistent and users might need control over this.  So long as its build scripts, the user can change names to control order. However, when we get to build script delegation, that won't work.

**Comment by epage** [comment]
Since this RFC was created...

We now have bindeps / artifact deps and I feel that pull in metabuild dependencies as artifact deps, rather than rlibs, would be better as it allows a shared build script to be compiled once.  Yes, the rlib gets compiled once but we then still have to have the wrapper pull it in and that can be significant, depending on the profile settings.

You can now `use dep::main;` to define a `main` function.

We have a request for being able to run unit tests (#9942).  If build scripts were dedicated packages in a workspace, then they would get that "for free".  Granted, this then requires publishing the build script package.

There is more scrutiny of supply chains and auditing of code that gets run during builds (e.g. #5720 is talked about frequently in the community).  This allowing us to only audit a common set of shared build scripts and not even wrapper scripts in user programs would be a big help.

There is a continued emphasis on build times and I would be concerned about this needing a toml parser, serde, and enough of the manifest schema to make this work.  That could very impede this proposal being adopted in core crates.  Yes, we could do intermediate solutions like pulling out `package.metadata`, serializing it to json, and passing that into the build script via an env variable, but that still requires `serde`, `serde_json`.

**Comment by epage** [comment]
A counter proposal to the current metabuild design, broken out into milestones

### Step 1: metabuild polyfill

Extend `package.build` to accept a list of build scripts
- The user can define build scripts that are just `use foo::main;`
- This also allows more purpose-built build scripts to audit, limit re-runs, etc

```toml
[package]
build = ["build-foo.rs", "build-bar.rs"]
```

This would allow some experimentation with this approach without fully defining the interface

*Note:* no auto-discovery of multiple build scripts is being considered because the end-goal is to shift the focus to metabuild

### Step 2: metabuild polyfill arguments

Extend `package.build` to define arguments and/or env variables to be passed to the build script

```toml
[[package.build]]
path = "build-foo.rs"
args = ["--foo=bar"]
env = { FOO = "bar" }
```

The focus is on providing a low-level mechanism that users can do what they want with without much compile-time overhead.

Ideally, args would be parsed with a CLI parser which can pull in some bulk.  Small parsers like `lexopt` help.   Custom test harnesses (rust-lang/testing-devex-team#2) will also need a parser, so if we encourage reuse between them than that will reduce the pain.

This would allow more build scripts to experiment with this approach

### Step 3: metabuild

**Blocked on artifact dependencies**

Add a `dependency` key to the `[[package.build]]` table
- `<dep>` must be a build-dependency that includes artifact type `bin`
- `<dep>` is skipped if the dependency is optional and not activated

```toml
[[package.build]]
dependency = "foo"
args = ["--foo=bar"]
env = { FOO = "bar" }
```

If the full table is not needed, maybe have a `package.build = "dep:foo"` shorthand

### Benefits

- We can gradually stabilize parts of this approach to allow experimenting with future parts
- Input to build scripts is tied to the build script definition
  - Organizationally nicer
  - Makes it clear that configuration is for a build script and not a random tool
  - Better draws the eye to something weird happening here **during** builds, making this less obfuscated
- Low compile-time overhead for a build scripts to meet the contractual obligations of this design
- Lower compile-time overhead by removing the need for wrapper build scripts

**Comment by joshtriplett** [comment]
I love the concept of this. Steps 1 and 3 together seem like a great design.

For step 2, I'm wondering if we need *both* env and args, particularly if cargo already passes all the environment variables it normally does to a build script.

Could we drop args and the fully general env support, and instead define a narrower config mechanism that we pass through via either the command line or the environment (one or the other, not both)?

For instance:

```toml
[[package.build]]
dependency = "foo"
config = { foo = "bar" }
```

We could either pass this through the environment as `CARGO_BUILD_CONFIG_foo="bar"`, or pass it on the command line as `foo=bar`. Either way, constraining this somewhat seems like it'd give a simpler result and clearer guidance, rather than a completely general channel for arguments and environment variables.

We could, alternatively, do the same thing for passing in the contents of (some subset of) the `metadata` table. The downside of doing *that* would be if people had substantial content in their metadata table already and didn't want or expect it to be passed to their build scripts. The advantage *and* disadvantage of using `metadata` that would be sharing it amongst build scripts; that would avoid duplication but require scripts to coordinate their interpretations of common elements.

On balance I think we should include it under `package.build`.

**Comment by epage** [comment]
I had both `env` and `args` as I was intentionally going very low level and decided to give the user all the knobs (well, except stdin).  I'm fine dropping `args`.

As for `env`, I had considered going fancier.  My questions are
- Is it worth the design trade off?
- How much do we need to design before moving forward?

I was hoping by going with `env` we could keep the design costs down and get this moving forward.

For benefits to a higher level config, I saw little.  I didn't see basic validation getting users much since the build script will validate anyways.  However, in thinking more on it, what we can get is
- Making the config a closed set, catching typos
- Potentially auto-generating documentation
- Some basic semver checking across releases

The main cost I thought of was users dealing with the translation of the config to env variables (prefixes, case conversion).

However, if we are to get those benefits I named, we also need to define a set of parameters to a build script.  We'd need a name for the table that has a clear role separate from `features` and whatever we call mutually exclusive features.  It might even be per-`[[bin]]`?  Maybe we can cheat and instead make the config be a field that uses the same syntax as `check-cfg`?  If so, we might want it to be an array of strings because merging all of those into one string could get messy.  Either way, would this format complexity be worth it for build script interface, an already low level mechanism that is a bit off in the weeds?

**Comment by tmandry** [comment]
I would push back on using env variables to pass configs: They are a global namespace, they make debugging failed commands much more difficult, and the experience if you mistype a name somewhere isn't good: often the command succeeds but silently does the wrong thing. We already have env variables that are global to the build/package, but for individual commands it makes less sense IMO.

As a start I would suggest that configs can be scalars that are translated into command line arguments using standard `--key=value` notation. Perhaps we could later support tables with `--key.nested1.nested2=value` notation or something else (it should be easy to support with a library like clap ideally.. although since these are build scripts, we ought to be careful about the impact on build time).

**Comment by epage** [comment]
If we have basic validation, I don't see how env variables are any worse than CLI arguments.  I would expect the validation from Cargo to be nearly as rich as what an optimum CLI can provide and can provide the feedback much earlier (before kicking off any builds).  The "delegate to a CLI" approach is used with lib`test` and it is really annoying to work with, e.g. having to wait for compilation to get any feedback.

I suspect build times (and annoyingly "dependency counts")  can make or break this feature and using a CLI could break this.  I'd like to see this used for nearly all build scripts and a lot of people, especially more core crates, are sensitive to these types of things.  In practice, it would help if they all used the same CLI parser but people rarely look at build time costs of dependencies in aggregate ("X pulls in `syn`" but you already have it from serde...).  You can use a [lighter weight CLI parser](https://github.com/rosetta-rs/argparse-rosetta-rs)  but then you either have to sacrifice rich errors or get a more convoluted API to work with.

**Comment by Kixunil** [comment]
Oh, yes please, metabuild using artifact dependencies makes so much sense! I think it'd even make sense to have a common cache of artifact dependencies since presumably tools don't need to change that often.

Regarding configuration I think env is fine and whoever needs something more structured can use `metadata`. @tmandry if a program is written to ignore missing env variable it will also ignore missing arguments.

One thing though, I think the tools should be categorized so that some can be skipped. E.g. codegen tools are required and would always run, link-info-providers would only run during linking and only if the information wasn't provided in other way (today this disables the entire build script which nukes bindgen), doc generators possibly shouldn't run by default.

**Comment by epage** [comment]
> One thing though, I think the tools should be categorized so that some can be skipped. E.g. codegen tools are required and would always run, link-info-providers would only run during linking and only if the information wasn't provided in other way (today this disables the entire build script which nukes bindgen), doc generators possibly shouldn't run by default.

In the short term, this would allow skipping through features.  In the longer term, by changing each build script into a table, we could add more features like this in the future.  I'd rather not commit to it now so we can focus on these more concrete steps and get it merged rather than getting side tracked with researching mew build script concepts.

**Comment by epage** [comment]
> The issue with features is that if one uses them for enabling/disabling link info they have to stay to not break things. I'd very much like if at least the link info thing was possible from the beginning because there's ton of crates doing FFI and also using bindgen.

Yes, it wouldn't work in a case like that but for things like doc gen.

**Comment by arielb1** [comment]
> If we have basic validation, I don't see how env variables are any worse than CLI arguments

The problem with "unscoped" env variables is that they allow for kinds of spooky actions (e.g. people using them to override `CC` rather than using some sort of "standard" method to override `CC`). If the env vars are scoped (as `CARGO_BUILD_<env>` etc.) they are less spooky.

[and if you intended to talk about scoped env vars all along, I couldn't tell that and probably some of the other issue readers couldn't as well]
